crates/mpc-protocols/src/mpc_service.rs

```diff
@@ -0,0 +1,115 @@
+use crate::mpc_events::{CreatedProofMPCEvent, MPCEvent};
+use pera_types::base_types::ObjectID;
+use pera_types::event::Event;
+use std::cell::RefCell;
+use std::collections::{HashMap, VecDeque};
+use tokio::sync::{mpsc, Mutex, RwLock};
+
+const MAX_ACTIVE_MPC_INSTANCES: usize = 100;
+
+pub enum MPCInput {
+    InitEvent(CreatedProofMPCEvent),
+    Message,
+    OutputEvent,
+}
+
+struct MPCInstance {
+    status: MPCStatus,
+    /// The channel to send message to this instance's message handler thread when this instance is active
+    messages_handler_sender: Option<mpsc::Sender<MPCInput>>,
+    pending_messages: Vec<MPCInput>,
+}
+
+/// Possible status of an MPC session
+/// - Active: The session is currently running, new messages will be forwarded to the session
+/// - Pending: The session is waiting for a slot to become active, when the init received there were already `MAX_ACTIVE_MPC_INSTANCES` active sessions
+/// - Finished: The session has finished, no more messages will be forwarded. The session will be removed from the service after a timeout.
+/// We want to keep it for some time so leftover messages related to it won't be treated as malicious.
+#[derive(Clone, Copy)]
+enum MPCStatus {
+    Active,
+    Pending,
+    Finished,
+}
+
+/// The `MPCService` is responsible for managing MPC instances:
+/// - It keeps track of all MPC instances
+/// - Runs the MPC session for each active instance
+/// - Ensures that the number of active sessions does not go over `MAX_ACTIVE_MPC_INSTANCES` at the same time
+pub struct MPCService {
+    mpc_instances: HashMap<ObjectID, MPCInstance>,
+    pending: VecDeque<ObjectID>,
+    active_instances_counter: usize,
```

Comment:
Keeping this counter updated is an error prone chore. Would it be possible to make this a computed property that filters `mpc_instances` to count the active instances?

Reply:
![image](https://github.com/user-attachments/assets/c5868153-b564-437b-8c51-c7b8e74613f2)

Reply:
Yes, please keep the counter. But we also have to ensure that the counter is not out of sync with the actual number of active instances (which I'm not convinced of yet). Perhaps this should be wrapped in a struct whose sole purpose is keeping track of that?

Reply:
Right now your'e right. I don't understand how a struct solves the issue; Anyhow, the way to go imo is:
1. When we implement instance completion, make sure we decrease the counter upon a completed instance (we'll have it's completed event).

2. Start an async task after any instance starts to change it's status after some reasonable timeout (10 minutes let's say) to "failed" and decrease the counter. 

Wdyt?

Reply:
I created an issue & added a todo

Reply:
Hmm, perhaps "wrapping [`mpc_instances`] with struct" is not the solution either. We just have to make sure that `mpc_instances` stays private.